### PR TITLE
Add option to get WebSession instance only if a session exists #26958

### LINKED ISSUE #26958: Add option to get WebSession instance only if a session exists
It would be nice if there were a way to request the session without creating an instance of one, similar conceptually to `HttpServletRequest#getSession(false)`.

Whenever the session is requested, say like so:

```java
exchange.getSession().filter((session) -> {
    if (session.getAttribute("attribute") != null) {
        // ...
    }
})
// ...
```

subscription will cause an `InMemoryWebSession` to be constructed. Since this is a blocking operation, it would be nice to avoid it when not needed.

If instead an application could do:

```java
exchange.getSessionOrEmpty().filter((session) -> {
    if (session.getAttribute("attribute") != null) {
        // ...
    }
})
```

then the predicate would not be invoked at all and construction of `InMemoryWebSession` would be avoided.

### PR DESCRIPTION
**Pull request by MaxBartkov**
<div class="js-timeline-item js-timeline-progressive-focus-container" data-gid="MDEyOklzc3VlQ29tbWVudDg0NTgyOTg4Ng==" style="box-sizing: border-box;"><div class="TimelineItem js-comment-container" data-gid="MDEyOklzc3VlQ29tbWVudDg0NTgyOTg4Ng==" data-url="/_render_node/MDEyOklzc3VlQ29tbWVudDg0NTgyOTg4Ng==/timeline/issue_comment?variables%5BdeferredCommentActions%5D=false" style="box-sizing: border-box; position: relative; display: flex; padding: 16px 0px; margin-left: 16px;"><div class=" timeline-comment-group js-minimizable-comment-group js-targetable-element TimelineItem-body my-0 " id="issuecomment-845829886" style="box-sizing: border-box; margin-top: 4px; margin-bottom: 0px !important; min-width: 0px; max-width: 100%; color: var(--color-timeline-text); flex: 1 1 auto;"><div class="ml-n3 timeline-comment unminimized-comment comment previewable-edit js-task-list-container editable-comment js-comment timeline-comment--caret reorderable-task-lists current-user" data-body-version="79fcd49d73e975d68827f251dca74efed7ebf7bb9df5af677682caea7e972026" data-unfurl-hide-url="/content_reference_attachments/hide" style="box-sizing: border-box; margin-left: -16px !important; position: relative; color: var(--color-text-primary); background-color: var(--color-bg-primary); border-color: var(--color-box-border-info); border-top-style: ; border-top-width: ; border-right-style: ; border-right-width: ; border-bottom-style: ; border-bottom-width: ; border-left-style: ; border-left-width: ; border-image-source: ; border-image-slice: ; border-image-width: ; border-image-outset: ; border-image-repeat: ; border-radius: 6px;"><div class="edit-comment-hide" style="box-sizing: border-box;"><task-lists sortable="" style="box-sizing: border-box;">the realized new method, it's connected with enhancement&nbsp;#26958 (comment)</task-lists></div><form class="js-comment-update" id="issuecomment-845829886-edit-form" action="https://github.com/spring-projects/spring-framework/issue_comments/845829886" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"></form></div></div></div></div><div class="js-timeline-item js-timeline-progressive-focus-container" data-gid="MDIwOkNyb3NzUmVmZXJlbmNlZEV2ZW50MTExNTY4MTEyMA==" style="box-sizing: border-box;"><div class="TimelineItem" style="box-sizing: border-box; position: relative; display: flex; padding: 16px 0px; margin-left: 16px;"><div class="TimelineItem-badge" style="box-sizing: border-box; position: relative; z-index: 1; display: flex; width: 32px; height: 32px; margin-right: 8px; margin-left: -15px; color: var(--color-icon-secondary); align-items: center; background-color: var(--color-timeline-badge-bg); border: 2px solid var(--color-bg-canvas); border-radius: 50%; justify-content: center; flex-shrink: 0; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><svg aria-hidden="true" viewBox="0 0 16 16" version="1.1" data-view-component="true" height="16" width="16" class="octicon octicon-cross-reference"><path fill-rule="evenodd" d="M16 1.25v4.146a.25.25 0 01-.427.177L14.03 4.03l-3.75 3.75a.75.75 0 11-1.06-1.06l3.75-3.75-1.543-1.543A.25.25 0 0111.604 1h4.146a.25.25 0 01.25.25zM2.75 3.5a.25.25 0 00-.25.25v7.5c0 .138.112.25.25.25h2a.75.75 0 01.75.75v2.19l2.72-2.72a.75.75 0 01.53-.22h4.5a.25.25 0 00.25-.25v-2.5a.75.75 0 111.5 0v2.5A1.75 1.75 0 0113.25 13H9.06l-2.573 2.573A1.457 1.457 0 014 14.543V13H2.75A1.75 1.75 0 011 11.25v-7.5C1 2.784 1.784 2 2.75 2h5.5a.75.75 0 010 1.5h-5.5z"></path></svg></div><div class="TimelineItem-body" style="box-sizing: border-box; min-width: 0px; max-width: 100%; margin-top: 4px; color: var(--color-timeline-text); flex: 1 1 auto; font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><br class="Apple-interchange-newline">

</div></div></div>

### CODE REVIEW

**Comment by MaxBartkov** [comment]
the realized new method, it's connected with enhancement https://github.com/spring-projects/spring-framework/issues/26958#issue-895570030

File: spring-web/src/main/java/org/springframework/web/server/ServerWebExchangeDecorator.java [hunk 1]
```diff
@@ -82,6 +82,14 @@ public ServerHttpResponse getResponse() {
 		return getDelegate().getSession();
 	}
 
+	@Override
+	public Mono<WebSession> getSessionOrEmpty() {
+		if (getDelegate().getSession() == null) {
+			return Mono.empty();
+		}
+		return getSession();
+	}
```
Comment (cuspymd) [inline spring-web/src/main/java/org/springframework/web/server/ServerWebExchangeDecorator.java:91]: All three implementations are the same. How about providing a default implementation in `ServerWebExchange`?

Comment (MaxBartkov) [reply spring-web/src/main/java/org/springframework/web/server/ServerWebExchangeDecorator.java:91]: All right with that. getSessionOrEmpty() using already created method getSession(). Difference between getSessionOrEmpty() and getSession() in checking for null

File: spring-web/src/main/java/org/springframework/web/server/adapter/DefaultServerWebExchange.java [hunk 2]
```diff
@@ -206,6 +206,14 @@ private HttpHeaders getResponseHeaders() {
 		return this.sessionMono;
 	}
 
+	@Override
+	public Mono<WebSession> getSessionOrEmpty() {
+		if (this.sessionMono == null) {
```
Comment (rstoyanchev) [inline spring-web/src/main/java/org/springframework/web/server/adapter/DefaultServerWebExchange.java:211]: `sessionMono` should never be null. It's initialized in the constructor.

**Review by rstoyanchev — COMMENTED** [review]
I don't see how this could work as proposed.

File: spring-web/src/main/java/org/springframework/web/server/ServerWebExchangeDecorator.java (hunk 1 above)
Comment (cuspymd) [reply spring-web/src/main/java/org/springframework/web/server/ServerWebExchangeDecorator.java:91]: I mean like below.
```
	default Mono<WebSession> getSessionOrEmpty() {
		if (getSession() == null) {
			return Mono.empty();
		}
		return getSession();
	}
```

**Comment by rstoyanchev** [comment]
This is not a straight-forward change that will involve changes to at least two contracts (ServerWebExchange and WebSessionManager) and there is a relationship between getSession and getSessionOrEmpty that is not as simple as checking for null, getSession will never return null. In short, I'm not sure we'll be able to make this change it in 5.3.x. For now I've added one commit to switch back to a "parallel" thread and that might be all we can do for the time being.

**Pull request closed without merging**